File: parse_gt3x.r
# 1. Paste files into files directory
# 2. CSV files will be created in specified output_directory (default: output)

library(read.gt3x)

list_files <- function(input_directory) {
    files <- list.files(paste(".", "files", sep = "/"))
    return(files)
}

transformENMO <- function(data, sample_frequency, window) {
    x <- data[, 2]
    y <- data[, 3]
    z <- data[, 4]
    ENMO <- sqrt(x^2 + y^2 + z^2) - 1
    ENMO[which(ENMO < 0)] <- 0
    ENMO2 <- cumsum(ENMO)
    ENMO3 <- diff(ENMO2[seq(1, length(ENMO), by = (sample_frequency * window))]) / (sample_frequency * window)

    return(ENMO3)
}

to_csv <- function(df, path, output_directory) {
    write.csv(df, paste(output_directory, path, sep = "/"))
}


parse_gt3x <- function(file, input_directory, output_directory) {
    split_string_v <- unlist(strsplit(file, "[.]"))
    split_filename <- split_string_v[-length(split_string_v)]

    data <- read.gt3x(paste(input_directory, file, sep = "/"))

    df <- as.data.frame(data)

    sample_rate <- attributes(df)[setdiff(names(attributes(df)), c("row.names"))]$sample_rate

    ENMO <- transformENMO(df, sample_rate, 1)

    to_csv(ENMO, paste(split_filename, ".csv", sep = ""), output_directory = output_directory)
}

main <- function(input_directory, output_directory) {
    dir.create(file.path(".", "output"), showWarnings = FALSE)

    files <- list_files(input_directory)

    sapply(files, parse_gt3x, input_directory = input_directory, output_directory = output_directory)
}

main("files", "output")

File: python/parse_gt3x.py
from pygt3x.reader import FileReader
from pygt3x.calibration import CalibratedReader
from math import sqrt

# Read raw data and calibrate
# Dump to pandas data frame
with FileReader("test.gt3x") as reader:
    dfdf = reader.to_pandas()
    calibrated_reader = CalibratedReader(reader)
    df = calibrated_reader.to_pandas()

    window = 1
    sample_frequency = 100

    df['ENMO'] = df['X']**2 + df['Y']**2 + df['Z']**2

    df['ENMO'] = df['ENMO'].apply(lambda x: sqrt(x) - 1)
    df.loc[df['ENMO'] < 0] = 0
    df['ENMO2'] = df['ENMO'].cumsum()
    
    df['ENMO2'].to_csv('test.csv')
    
    sample_rate = window * sample_frequency

    df = df.iloc[1::sample_rate, :].reset_index()

    df = df['ENMO2'].diff() / 100

    df.to_csv('test.csv')